Utils/Dataflow/017_adjustMetadata/adjustMetadata.py

```diff
@@ -170,23 +175,41 @@ def transform_chain_data(data):
                          'non-dict data: %s. Skipping.\n' % str(data))
         return False
     chain_data = data.get('chain_data')
-    if not chain_data or not chain_data.replace(',', '').isdigit():
+    if not chain_data:
         taskid = data.get('taskid')
         if not taskid:
             sys.stderr.write('(WARN) Function transform_chain_data() '
-                             'received data with incorrect chain_data and '
+                             'received data with empty chain_data and '
                              'without taskid: %s. Skipping.\n' % str(data))
             return False
-        sys.stderr.write('(WARN) Task %s: cannot transform chain_data "%s", '
-                         'it seems to be incorrect. Setting chain_id=%s, '
-                         'chain_data=[%s].\n'
+        sys.stderr.write('(WARN) Task %s: chain_data field is empty.'
+                         'Setting chain_id=%s, chain_data=[%s].\n'
                          % (taskid, chain_data, taskid, taskid))
         data['chain_id'] = taskid
         data['chain_data'] = [taskid]
         return False
-    chain_data = [int(i) for i in chain_data.split(',')]
-    data['chain_id'] = chain_data[0]
-    data['chain_data'] = chain_data
+    chain_items = chain_data.split(',')
+    try:
+        parent_id, parent_name, parent_events = chain_items[-2].split(':')
+        chain_items[-2] = parent_id
+        data['parent_taskname'] = parent_name
+        data['parent_total_events'] = parent_events
+    except IndexError:
+        # No parent (chain root)
+        pass
+    except ValueError:
+        # Invalid parent data
+        sys.stderr.write('(WARN) Invalid parent task data (expected format:'
+                         ' "p_tid:p_name:p_total_events"): %s (tid: %s).\n'
+                         % (chain_items[-2], data.get('taskid')))
+    try:
+        chain_data = [int(i) for i in chain_items]
+        data['chain_id'] = chain_data[0]
+        data['chain_data'] = chain_data
+    except ValueError, err:
+        sys.stderr.write('(WARN) Invalid chain_data item: %s (tid: %s).\n'
+                         % (err, data.get('taskid')))
+        return False
```

Comment:
Before this PR, correct `taskid` + invalid `chain_data` caused the task to be defined as a root of its own chain.

```
>>> d = {"chain_data":"1,2B,3", "taskid":10}
>>> adjustMetadata.transform_chain_data(d)
(WARN) Task 10: cannot transform chain_data "1,2B,3", it seems to be incorrect. Setting chain_id=10, chain_data=[10].
False
>>> d
{'chain_data': [10], 'chain_id': 10, 'taskid': 10}
```

Now, since the check in former line 173 (`if not chain_data or ...`) was changed, the same situation will be handled in this exception, and `chain_id` and `chain_data` fields will not be set/updated at all.

```
>>> d = {"chain_data":"1,2B,3:NAME:55,4", "taskid":10}
>>> adjustMetadata.transform_chain_data(d)
(WARN) Invalid chain_data item: invalid literal for int() with base 10: '2B' (tid: 10).
False
>>> d
{'parent_taskname': 'NAME', 'chain_data': '1,2B,3:NAME:55,4', 'parent_total_events': '55', 'taskid': 10}
```

P.S. Yes, I know that `taskid` of `10` makes no sense since it's not present in the `chain_data`.

Reply:
Right, the behaviour has changed -- and most likely ES will be surprised by the string `chain_data`, so there're at least two reasons to do something about it.

But I have a question. I am not sure that it's *correct* to set `chain_data` and `chain_id` to `taskid` when original `chain_data` is so wierd -- I'd rather leave them empty. What do you think? Of course, in this exact PR this change is out of place -- it was unintended and I won't even try to keep it here -- but it may be forgotten for good or just moved to a separate PR, so I'd like to know your opinion.

Reply:
And, talking about it: in original logic in case of missed `taskid` the function would still leave `chain_data` as is:
```
~$ echo '{"chain_data":"1,2B,3", "taskid":null}' | ./adjustMetadata.py -m s -E ''
<...>
2020-06-24 15:29:08 (INFO) (ProcessorStage) Starting stage execution.
(WARN) Function transform_chain_data() received data with incorrect chain_data and without taskid: {u'chain_data': u'1,2B,3', 'phys_category': ['Uncategorized'], u'taskid': None}. Skipping.
{"chain_data": "1,2B,3", "taskid": null, "phys_category": ["Uncategorized"]}
2020-06-24 15:29:08 (INFO) (ProcessorStage) Stopping stage.
```
-- which is also a questionable behaviour; I mean, the stage is supposed to set `chain_data` to an expected format -- list of task IDs -- which it does not; so any post-processing relying on this will fail. Setting it to `null` would look fairer to me.
*(Also, out of this PR, but since we're here...)*

Reply:
Original logic restored: 0d9c09a
```
(dkb-dev) [dkb@aiatlas171 017_adjustMetadata]$ echo '{"chain_data":"1,2B,3:NAME:55,4", "taskid":10}' | ./adjustMetadata.py -m s -E ''
<...>
2020-06-24 15:40:09 (INFO) (ProcessorStage) Starting stage execution.
(WARN) Invalid chain_data item: invalid literal for int() with base 10: '2B' (tid: 10).
(INFO) Setting chain_id=10, chain_data=[10] (tid: 10).
{"parent_total_events": "55", "chain_data": [10], "chain_id": 10, "phys_category": ["Uncategorized"], "taskid": 10, "parent_taskname": "NAME"}
2020-06-24 15:40:09 (INFO) (ProcessorStage) Stopping stage.
```

Reply:
> But I have a question. I am not sure that it's _correct_ to set `chain_data` and `chain_id` to `taskid` when original `chain_data` is so wierd -- I'd rather leave them empty. What do you think? Of course, in this exact PR this change is out of place -- it was unintended and I won't even try to keep it here -- but it may be forgotten for good or just moved to a separate PR, so I'd like to know your opinion.

My opinion is as following: you are asking reasonable questions, both here and in other parts of your replies, and I had similar thoughts while reviewing this code. Since possible changes are outside of this PR's code, I made a [card](https://trello.com/c/rLxtg5Mn) for discussing the issues - will write some more thoughts into it later.

As for this point here, "things are now as they were before", so we are concluding it.